service/storage_service.cc

```diff
@@ -2497,6 +2497,92 @@ canonical_mutation storage_service::build_mutation_from_join_params(const join_n
     return builder.build();
 }
 
+class join_node_rpc_handshaker : public service::group0_handshaker {
+private:
+    service::storage_service& _ss;
+    const join_node_request_params& _req;
+
+public:
+    join_node_rpc_handshaker(service::storage_service& ss, const join_node_request_params& req)
+            : _ss(ss)
+            , _req(req)
+    {}
+
+    future<bool> perform_handshake(const group0_info& g0_info) override {
+        return _ss.raft_perform_join_handshake(g0_info, _req);
+    }
+};
+
+future<bool> storage_service::raft_perform_join_handshake(const group0_info& g0_info, const join_node_request_params& params) {
+    slogger.info("raft topology: join: sending the join request to {}", g0_info.ip_addr);
+
+    auto result = co_await ser::join_node_rpc_verbs::send_join_node_request(
+            &_messaging.local(), netw::msg_addr(g0_info.ip_addr), g0_info.id, std::move(params));
+    co_return co_await std::visit(overloaded_functor {
+        [this] (const join_node_request_result::ok&) -> future<bool> {
+            // The request has been placed to system.topology.
+            // Now we need to wait until the topology coordinator
+            // either accepts us or rejects us.
+            slogger.info("raft topology: join: request to join placed, waiting  "
+                         "for the response from the topology coordinator");
+
+            _join_node_request_done.set_value();
+
+            // Processing of the response is done in `join_node_response_handler`.
+            // Wait for it to complete.
+            // TODO: Timeouts?
+            co_await _join_node_result.get_shared_future(_abort_source);
```

Comment:
Well, since the response may get lost we need the timeout.

Reply:
Now there is a timeout (30s).